src/main/java/org/testng/internal/XmlMethodSelector.java

```diff
@@ -142,14 +144,19 @@ else if (isExcludedInGroups) {
 
           List<String> includedMethods =
               createQualifiedMethodNames(xmlClass, toStringList(xmlClass.getIncludedMethods()));
-          boolean isIncludedInMethods = isIncluded(fullyQualifiedMethodName, includedMethods);
+          boolean isIncludedInMethods = isIncluded(fullyQualifiedMethodName, includedMethods, noGroupsSpecified);
           List<String> excludedMethods = createQualifiedMethodNames(xmlClass,
               xmlClass.getExcludedMethods());
           boolean isExcludedInMethods = isExcluded(fullyQualifiedMethodName, excludedMethods);
           if (result) {
             // If we're about to include this method by group, make sure
             // it's included by method and not excluded by method
-            result = isIncludedInMethods && ! isExcludedInMethods;
+            if (!xmlClass.getIncludedMethods().isEmpty()) {
```

Comment:
The new `isIncluded()` works on methods that were filtered via `createQualifiedMethodNames(xmlClass, toStringList(xmlClass.getIncludedMethods()));`. But we need to involve the `isIncludedInMethods` and `isIncludedInMethods` if and only if there were some `<include>` or `<exclude>` methods specified by the users. That is why this check is required.